cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_service_kernels_def_dpc.hpp

```diff
@@ -0,0 +1,926 @@
+                        const Index pos_new =
+                            (to_right ? left_ch_row_count + group_right_boundary + boundary
+                                      : group_left_boundary + i - ind_start - boundary);
+                        tree_order_buf_ptr[offset + pos_new] = id;
+                        group_row_to_right_count += reduce(sbg, to_right, plus<Index>());
+                    }
+                }
+            }
+        });
+    });
+
+    event.wait_and_throw();
+
+    std::swap(tree_order, tree_order_buf);
+    return event;
+}
+
+template <typename Float, typename Bin, typename Index, typename Task>
+sycl::event train_service_kernels<Float, Bin, Index, Task>::update_mdi_var_importance(
+    const pr::ndarray<Index, 1>& node_list,
+    const pr::ndarray<Float, 1>& node_imp_decrease_list,
+    pr::ndarray<Float, 1>& res_var_imp,
+    Index data_column_count,
+    Index node_count,
+    const be::event_vector& deps) {
+    ONEDAL_ASSERT(node_list.get_count() == node_count * impl_const_t::node_prop_count_);
+    ONEDAL_ASSERT(node_imp_decrease_list.get_count() == node_count);
+    ONEDAL_ASSERT(res_var_imp.get_count() == data_column_count);
+
+    Index local_size = preferable_group_size_;
+    //calculating local size in way to have all subgroups for node in one group to use local buffer
+    while (local_size > node_count && local_size > preferable_sbg_size_) {
+        local_size >>= 1;
+    }
+
+    const Index* node_list_ptr = node_list.get_data();
+    const Float* node_imp_decrease_list_ptr = node_imp_decrease_list.get_data();
+    Float* res_var_imp_ptr = res_var_imp.get_mutable_data();
+
+    const sycl::nd_range<2> nd_range =
+        be::make_multiple_nd_range_2d({ local_size, data_column_count }, { local_size, 1 });
+
+    const Index node_prop_count =
+        impl_const_t::node_prop_count_; // num of split attributes for node
+    const Index leaf_mark = impl_const_t::leaf_mark_;
+    const Index max_sub_groups_num =
+        max_sbg_count_per_group_; //need to calculate it via device info
+
+    auto event = queue_.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        sycl::accessor<Float, 1, sycl::access::mode::read_write, sycl::access::target::local> buf(
+            max_sub_groups_num,
+            cgh);
+
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<2> item) {
+            auto sbg = item.get_sub_group();
+
+            const Index local_id = item.get_local_id()[0];
+            const Index sub_group_local_id = sbg.get_local_id();
+            const Index sub_group_size = sbg.get_local_range()[0];
+            const Index local_size = item.get_local_range()[0];
+            const Index n_sub_groups =
+                local_size / sub_group_size; // num of subgroups for current node processing
+            const Index sub_group_id = local_id / sub_group_size;
+
+            const Index buf_idx =
+                item.get_global_id()[1] %
```

Comment:
```suggestion
                item.get_global_id(1) %
```

Reply:
i made it for consistency because for example sbg.get_local_range()[0] can't be replaced with sbg.get_local_range(0), therefore to make in one style i use [] for all.